src/core/ext/filters/client_channel/xds/xds_bootstrap.cc

```diff
@@ -103,9 +103,33 @@ XdsBootstrap::~XdsBootstrap() {
   grpc_slice_unref_internal(contents_);
 }
 
-grpc_error* XdsBootstrap::ParseXdsServer(grpc_json* json) {
+grpc_error* XdsBootstrap::ParseXdsServerList(grpc_json* json) {
   InlinedVector<grpc_error*, 1> error_list;
-  server_uri_ = nullptr;
+  size_t idx = 0;
+  for (grpc_json *child = json->child; child != nullptr;
+       child = child->next, ++idx) {
+    if (child->key != nullptr) {
+      char* msg;
+      gpr_asprintf(&msg, "array element %" PRIuPTR " key is not null", idx);
+      error_list.push_back(GRPC_ERROR_CREATE_FROM_COPIED_STRING(msg));
```

Comment:
This looks like an impossible case here and would indicate a bug. Other objects in the json may be corrupted. Maybe assert? Or at least not to proceed?

Reply:
It should be impossible, yes.  But in general, we're trying to avoid assertions in our code, so it's better to handle the error case without one, even if we never expect it to happen.

In actuality, the fact that we have to check this at all is actually a symptom of the fact that our current JSON API is not very well suited to what we need to do with it.  I am working on a series of PRs that will add a new JSON API, and then I'll convert this code to use it, which should clean this up quite a bit.

Reply:
In general I don't think assertion is a bad thing; maybe we should have an assertion that's effective in Debug mode but not in Release mode so we can feel free to use it more. I'm not going to push hard though.

I usually find it easier to debug when we assert rather than letting things go on, so that when such a bug really occurs, it is easy to find out that the bug comes from JSON instead of needing to trace back the logics.

Reply:
It's nice to hear there'll be a new JSON API. I also found it confusing when reading the JSON usage :smiley: